### PR TITLE
Add some migration helper to help migrating pallet changing pallet prefix

### PR DESCRIPTION
**Pull request by thiolliere**
when migrating from decl_storage to frame v2, some pallet needs to migrate their storage too.

This is some helper to move some specific storage in case the pallet prefix is shared by multiple pallets (e.g. treasury and bounty) or to move all pallet.

This can be useful for
https://github.com/paritytech/substrate/pull/8193
https://github.com/paritytech/substrate/pull/8196
https://github.com/paritytech/substrate/pull/8044

### CODE REVIEW

File: frame/support/src/storage/migration.rs [hunk 1]
```diff
@@ -195,3 +197,134 @@ pub fn take_storage_item<K: Encode + Sized, T: Decode + Sized, H: StorageHasher>
 ) -> Option<T> {
 	take_storage_value(module, item, key.using_encoded(H::hash).as_ref())
 }
+
+/// Move a storage from a pallet prefix to another pallet prefix.
+pub fn move_storage_from_pallet(item: &[u8], old_pallet: &[u8], new_pallet: &[u8]) {
```
Comment (thiolliere) [inline frame/support/src/storage/migration.rs:202]: this can't work for storage map :-/, I'm on it

Comment (thiolliere) [reply frame/support/src/storage/migration.rs:205]: fixed with test

**Comment by thiolliere** [comment]
I'm not sure who to ping for review, but those should be easy to review and are helping for migrating to framev2

File: frame/support/src/storage/migration.rs [hunk 2]
```diff
@@ -195,3 +197,154 @@ pub fn take_storage_item<K: Encode + Sized, T: Decode + Sized, H: StorageHasher>
...
+			OldStorageMap::insert(3, 4);
+
+			move_storage_from_pallet(b"foo_map", b"my_old_pallet", b"my_new_pallet");
+
+			assert_eq!(OldStorageValue::get(), Some(3));
+			assert_eq!(OldStorageMap::iter().collect::<Vec<_>>(), vec![]);
+			assert_eq!(NewStorageValue::get(), None);
+			assert_eq!(NewStorageMap::iter().collect::<Vec<_>>(), vec![(1, 2), (3, 4)]);
+
+			move_storage_from_pallet(b"foo_value", b"my_old_pallet", b"my_new_pallet");
```
Comment (apopiak) [inline frame/support/src/storage/migration.rs:326]: This is cool, but would it be possible to allow specifying the pallet info (as well)?
Would be nice if we could avoid bare strings where possible

Comment (thiolliere) [reply frame/support/src/storage/migration.rs:326]: yes we could do:

```
fn migrate_::<PI: PalletInfo, P: Pallet: 'static>(storage: &str, from_pallet: &str) {
	// move to `PalletInfo::name::<P>()`
}
```

But I thought caller could just do:
```
move_storage_from_pallet(b"foo", b"old_prefix", T::PalletInfo::name::<Pallet>().as_bytes())
```

File: frame/support/src/storage/migration.rs [hunk 3]
```diff
@@ -195,3 +197,154 @@ pub fn take_storage_item<K: Encode + Sized, T: Decode + Sized, H: StorageHasher>
...
 	take_storage_value(module, item, key.using_encoded(H::hash).as_ref())
 }
+
+/// Move a storage from a pallet prefix to another pallet prefix.
+///
+/// NOTE: It actually moves every key after the storage prefix to the new storage prefix,
+/// key on storage prefix is also moved.
+pub fn move_storage_from_pallet(item: &[u8], old_pallet: &[u8], new_pallet: &[u8]) {
+	let mut new_prefix = Vec::new();
+	new_prefix.extend_from_slice(&Twox128::hash(new_pallet));
```
Comment (kianenigma) [inline frame/support/src/storage/migration.rs:207]: can we reuse anything from `StorageValue` struct to generate this key? I am worried that we have hardcoded this pattern of `twox ++ twox` in many places. External teams have probably made this mistake even more.

**Review by kianenigma — APPROVED** [review]
I was actually waiting for something like this before finishing #8044!

File: frame/support/src/storage/migration.rs (hunk 3 above)
Comment (thiolliere) [reply frame/support/src/storage/migration.rs:207]: unfortunately I made the generator traits StorageValue, StorageMap and StorageDoubleMap abstract on the key generation with a default implementation :-/ so we can't make use of it directly.
We can still introduce a function for generating this key though.

Also the end I think we can refactor quite a bit if we get rid of decl_storage.

File: frame/support/src/storage/migration.rs (hunk 2 above)
Comment (thiolliere) [reply frame/support/src/storage/migration.rs:326]: the migration function in this module are loosely typed. I think it offers good flexibility so I think it is fine.
Like if someone want to migrate from old macro to new macro, old macro to old macro with another prefix, or new macro to new macro, then they can make use of this simple function.

File: frame/support/src/storage/migration.rs [hunk 4]
```diff
@@ -195,3 +197,154 @@ pub fn take_storage_item<K: Encode + Sized, T: Decode + Sized, H: StorageHasher>
...
+
+	let iter = PrefixIterator {
+		prefix: from_prefix.to_vec(),
+		previous_key: from_prefix.to_vec(),
+		drain: true,
+		closure: |key, value| Ok((key.to_vec(), value.to_vec())),
+	};
+
+	for (key, value) in iter {
+		let full_key = [to_prefix.clone(), &key].concat();
```
Comment (bkchr) [inline frame/support/src/storage/migration.rs:246]:
```suggestion
		let full_key = [to_prefix, &key].concat();
```

File: frame/support/src/storage/migration.rs [hunk 5]
```diff
@@ -195,3 +197,154 @@ pub fn take_storage_item<K: Encode + Sized, T: Decode + Sized, H: StorageHasher>
 ) -> Option<T> {
 	take_storage_value(module, item, key.using_encoded(H::hash).as_ref())
 }
+
+/// Move a storage from a pallet prefix to another pallet prefix.
+///
+/// NOTE: It actually moves every key after the storage prefix to the new storage prefix,
+/// key on storage prefix is also moved.
+pub fn move_storage_from_pallet(item: &[u8], old_pallet: &[u8], new_pallet: &[u8]) {
```
Comment (bkchr) [inline frame/support/src/storage/migration.rs:205]: Maybe you could add some example and more docs on what these parameters are?

This is relative complicated to understand from the current docs.

File: frame/support/src/storage/migration.rs [hunk 6]
```diff
@@ -195,3 +197,154 @@ pub fn take_storage_item<K: Encode + Sized, T: Decode + Sized, H: StorageHasher>
...
+		unhashed::put_raw(&new_prefix, &value);
+		unhashed::kill(&old_prefix);
+	}
+}
+
+/// Move all storage key after the pallet prefix `old_pallet` to `new_pallet`
+///
+/// NOTE: It actually moves every key after the pallet prefix to the new pallet prefix,
+/// key on pallet prefix is not moved.
+pub fn move_pallet(old_pallet: &[u8], new_pallet: &[u8]) {
```
Comment (bkchr) [inline frame/support/src/storage/migration.rs:226]: Same

**Review by bkchr — APPROVED** [review]
Looks good, but docs need to be improved.

All these `from_prefix` and `to_prefix` are sometimes hashed internally, sometimes not. Would be nice to get better docs on this.

File: frame/support/src/storage/migration.rs (hunk 5 above)
Comment (thiolliere) [reply frame/support/src/storage/migration.rs]: indeed, I improved, I hope its clearer now.

**Review by bkchr — APPROVED** [review]

**Pull request merged**